### PR TITLE
TIG-1201 Genny: Create Quiesce System Utility Method

### PR DESCRIPTION
**Pull request by aggrand**
Implements topology discovery, and the quiesce method which waits for the oplog to catch up, fsyncs, and waits for collection to drop.

Also took a stab at formalizing the topology traversal code. I feel like in e.g. resmoke we end up with every kitchen-sink method attached to the cluster representation, whereas the shell seems to have ad-hoc traversal everywhere, so I tried to get around both of those.

The actual quiescing followed the JS implementation closely, so I didn't add automated testing for that part as discussed in the ticket and templatizing on both `client` and `database` and `collection` seems tricky, but I did check locally that it works for different clusters. It could be possible to wrap some of those interactions in `DBService` methods to better test it though, if you think that'd be worth doing. Or, after TIG-2823, doing a resmoke integration test.

### CODE REVIEW

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [inline src/gennylib/include/gennylib/quiesce.hpp:97]: This logic is different for 4.0 and earlier, but my understanding is that we use legacy genny on the legacy DSI branches, so I assumed adding a version check here would be unnecessary?

File: src/gennylib/include/gennylib/service.hpp
Comment (aggrand) [inline src/gennylib/include/gennylib/service.hpp:1]: I added this file because it made unit testing `Topology` easier. Another option would be to templatize it but I wanted to avoid having all the details of discovery in the header file.

File: src/cast_core/src/QuiesceActor.cpp
Comment (rtimmons) [inline src/cast_core/src/QuiesceActor.cpp:42]: nit but you don't need this error-checking - `Node` does it for you when you call `.to<>()`:

```
PhaseConfig(ctx)
: dbName{context.actor()["Database"].to<string>()} {}
```

will barf with a helpful error message if not present.

File: src/cast_core/src/QuiesceActor.cpp
Comment (rtimmons) [inline src/cast_core/src/QuiesceActor.cpp:54]: note to circle back here - it seems odd to require a db name to quiesce teh whole cluster

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:36]: 1. see if it makes sense to put most/all of these free functions into `topology.hpp`?
2. these aren't templated so you probably only need to decl them and then define them in the cpp (either topology or quiesce cpp). Not a huge deal but potentially speeds up some compile and link etc

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:52]: what other states could we be? standalone?

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:69]: I suspect you're just copying the js impl here and that there's a reason for the `{x:flush}` doc and whatnot, but this seems like voodoo. Can you either add a comment about what this is doing from a conceptual level or add a permalink to the js impl?

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:73]: 💅🏽 
1. `camelCase`
2. Please put this in `private` and with leading underscore
3. I don't grok why this is stateful or what this `success_acc` actually means
4. If we run this member-function a second time do we need to re-run the state-check loop above again? If so, do we want to reset `success_acc` to false? (Probably a dumb question given I don't quite grok what this var is doing)

This took me a while to parse as a member-variable 😸

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [reply src/gennylib/include/gennylib/quiesce.hpp:97]: The `legacy` branch is going away so all branches will be on genny master. It's not unreasonable to say that workloads that use this can't be run on 4.0. If anyone balks we can perhaps do runtime version checks.

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:94]: note to self that `v` is stack-allocated: ensure pass-by-lvalue isn't allowed

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:107]: I'm not sure if you need any of these helpers to be "visible" once you move the impls to cpp - perhaps just need to decl the `quiesce` free-function and the rest can just live in the anonymous namespace in the cpp.

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/quiesce.hpp:113]: 1. I'm super skeptical of how this will play with `PhaseLoop`. I'm holding my breath for now but note to self to come back here.
2. separate thought: do we want to ensure there aren't concurrent invocations of this?

File: src/gennylib/include/gennylib/service.hpp
Comment (rtimmons) [reply src/gennylib/include/gennylib/service.hpp:1]: Note to self to come back here.

I think this is a good design. Genny doesn't use a lot of virtuals in its core code but virtuals are a great strategy to ease testing in utilities.

But `service` is a bit of a smurf term. Perhaps it makes sense to rename these and make them have a more concrete connection (both in name and code location) to `topology.hpp`.

File: src/gennylib/include/gennylib/service.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/service.hpp:59]: may as well make everything virtual and make the impl in the cpp

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:29]: super nit: Just call this `Topology` - the fact that it's abstract is its business, and "description" doesn't add much to the name. Push back if you disagree no biggie.

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:56]: Consider names

1. `on{Before,After}Topology`
2. `on{Before,After}{Mongo{d,s},Shard}`
3. `onBetween{Mongod,Shard}s`

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:72]: 1. Why not pure-virtual (`= 0`)?
2. If most visitors won't need them maybe just don't include them for now?

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:74]: let's make this `=0` to make this a pure virtual class.

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:80]: sane here - `=0`.

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:214]: impls can be in .cpp (I realize it's annoying so my apologies for asking for banal code ceremony)

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:168]: likely wants to be private

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:175]: erm. seems odd for `class Topology` to have a `_topology` member-var.

File: src/gennylib/include/gennylib/topology.hpp
Comment (rtimmons) [inline src/gennylib/include/gennylib/topology.hpp:178]: hey this is neat

File: src/gennylib/src/topology.cpp
Comment (rtimmons) [inline src/gennylib/src/topology.cpp:37]: maybe make the uri a ctor arg of the `MongoDescription`? (there are a few other things here that maybe want to be pushed into impl details (perhaps in ctors) of the `*Description` classes. Not sure of implications on testing but may actually make it easier so not a strong opinion here yet.

File: src/gennylib/src/topology.cpp
Comment (rtimmons) [inline src/gennylib/src/topology.cpp:70]: I think this would be more testable if you had a `MongodDescription` ctor that took in a `view_or_value` (which you can trivially construct without live connections to a db).

File: src/gennylib/src/topology.cpp
Comment (rtimmons) [inline src/gennylib/src/topology.cpp:34]: odd to have a `void` `get` function - maybe this wants to be just a helper/free function or maybe `s/get/compute/`.

**Review by rtimmons — COMMENTED** [review]
Nice starting point and clean code. I don't know the js impl well (er..even at all) so I'm going to trust your faithfulness to that impl.

I didn't quite finish yet - just started poking at the cpp files and tests - but putting what I have so far. Take a quick look for now and let me know if I've gone down the wrong path on what to look at. Otherwise I'll poke more at the impl this afternoon.

File: src/cast_core/src/QuiesceActor.cpp
Comment (aggrand) [reply src/cast_core/src/QuiesceActor.cpp:54]: This is because my understanding is the collection drop checking is db-specific. At least in the 4.0-and-earlier implementation in the js it was. The new implementation uses `serverStatus` which [says](https://docs.mongodb.com/manual/reference/command/serverStatus/) it provides an overview of the db's state so I assume this added field is also db-specific.

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/quiesce.hpp:52]: There are a few [exotic](https://docs.mongodb.com/manual/reference/replica-states/) states if the replica set isn't in a stable position. This check was just copied from the js though we can delete if we don't think genny will ever talk to a repl set in those states.

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/quiesce.hpp:69]: Yeah will do! The js code isn't explained either but my understanding is to force the oplog to flush by doing a write with `w: all` and `j: true` so when that write is finished you know everything before it is replicated.

There's also [awaitReplication()](https://github.com/mongodb/mongo/blob/r4.1.4/src/mongo/shell/replsettest.js#L1354) in the shell which I think does something similar but its implementation is _way_ more complicated.

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/quiesce.hpp:73]: Will make those changes. It's stateful because in a sharded cluster `visitReplSetPre` will be called in the traversal on each shard, so we need to only return a success if the op succeeded on every shard.

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:72]: 1. The visit methods can't be pure-virtual because they're always called in the topology description traversal, even as a nop. The idea is that subclass visitors only need to override the visit methods they care about, instead of overriding to add a bunch of nops.
2. The `JsonVisitor` needs them. That visitor isn't 100% necessary but it was useful for testing both the discovery and that traversal works correctly.

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:74]: I thought pure virtual destructors were bad practice? Because the linker requires a definition for the destructor whether it's pure virtual or not, since every destructor call also calls the parent's destructor, its parent's, etc so a definition would need to be added outside the class declaration and having at least one function be pure virtual is all that's needed to make the class an abstract class?

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/quiesce.hpp:73]: I made those changes but as a more general question, do we care a lot about those standards (private, etc) for function-level classes like this? This object is sort of serving as a lambda

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/quiesce.hpp:107]: Yeah moved!

File: src/gennylib/include/gennylib/service.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/service.hpp:1]: I renamed it to `Connection`. Part of why I put it in a separate file is I thought it might be useful elsewhere in other places, the only other way we have to test driver-level things is with the resmoke integration and APM stuff, which might be more involved. Let me know though if we should just cross that bridge when we get there and I'll put it in the `topology.hpp` file.

File: src/gennylib/src/topology.cpp
Comment (aggrand) [reply src/gennylib/src/topology.cpp:70]: I tried this, but unfortunately a bson `element` doesn't seem to be a `view`, so couldn't be passed in, unless I'm overlooking something. I tried doing two different constructors, one for `element` and one for `std::string`, but it turns out that the document `element` and the `array::element` are different. They share a private base class that I can't seem to figure out how to convince the compiler to let me `dynamic_cast` to.

I figured having three different implementations of that constructor might not be an improvement over the existing situation, but let me know if that'd still be better!

File: src/cast_core/src/QuiesceActor.cpp
Comment (aggrand) [reply src/cast_core/src/QuiesceActor.cpp:54]: An alternative might be to discover all the dbs and then wait for drops on them?

File: src/gennylib/include/gennylib/quiesce.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/quiesce.hpp]: Added a locking mechanism to `quiesce.cpp` for the latter

**Review by rtimmons — APPROVED** [review]
Approved modulo some moving of things around and doing a bit of syntax 💅🏽 . Please push back wherever you disagree.

1. I looked at the actual quiesce business-logic and it seems reasonable, but please do put a link somewhere to the js impl for future maintainers. 
2. On a similar, note, consider asking Max H (`@visemet`) if he wouldn't mind taking a look at the quiesce cpp. He may also have good insights on the other C++isms, but he has a pretty good understanding of the quiesce logic iirc. Getting a quick 👍🏽 from him--if he has time--would be great.
3. Let's consider marking this as "beta" for now, even if just in a comment. I have very little doubt that this is correct (it's certainly well-designed and has reasonable testing), but I'd like us to know about uses of this so we can keep an eye on them. We're quite likely to be asked to field BFs if there is any suspicion of the quiesce logic making things go pear-shaped, so if we can proactively understand where it's being used we can double-check our impl and their usage is correct.

File: src/gennylib/src/quiesce.cpp
Comment (aggrand) [reply src/gennylib/src/quiesce.cpp:141]: I wasn't too sure what we wanted the concurrent behavior to be here. Really I think it only makes sense to ever have one, single-threaded actor do these quiesces. But if this function was ever incorporated into a larger actor or something, I think we want:

1. Only one thread does the quiesce.
2. The function only exits successfully when a quiesce has just occurred.
3. A thread won't do a redundant quiesce.

So this implementation lets one thread quiesce, and if other threads enter the function they fail to get the lock immediately, so block waiting for the lock. When they get it, they know the quiesce just happened so they can unlock and leave.

Something similar could be done with condition variables, but we'd have to deal with spurious wakeups and such so this seemed simpler and comparably efficient.

Comment (aggrand) [reply src/gennylib/src/quiesce.cpp:141]: I'll add a comment explaining this.

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:116]: I was kind of thinking having these in the header would be nice for publicly documenting the traversal order of the cluster, but most of the time that's probably not a major concern so I'm good with moving it.

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:124]: This isn't possible with the `onBetweenShards`, since we don't want it called after the last iteration. Maybe a bit of an unnecessary visit method, but `ToJsonVisitor` used it and it was nice for debugging.

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:66]: `onAfterSharded` refers to a sharded cluster, whereas `onAfterShards` is after going through each shard in the shard list. I'll rename `onAfterSharded` to `onAfterShardedCluster` to clarify.

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:116]: I think this would preclude visitors that accumulate state as they traverse the cluster, like `ToJsonVisitor`

File: src/gennylib/include/gennylib/topology.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/topology.hpp:168]: It should be safe to pass-by-reference since `update` doesn't store `DBConnection` anywhere and the enclosing scope outlives the `update` function? My concern with pass-by-value is that we don't want to create a new connection to the server, we want to reuse the existing one. And I think we don't want to have the `DBConnection` use `shard_ptr` to store the `mongocxx::client`?

File: src/gennylib/include/gennylib/Connection.hpp
Comment (aggrand) [reply src/gennylib/include/gennylib/Connection.hpp:48]: Thanks this is amazing I added it everywhere.

File: src/gennylib/src/topology.cpp
Comment (aggrand) [reply src/gennylib/src/topology.cpp:32]: This is interesting. Some of the names _do_ have the form `hostname:port` without the '/', but this code was still working fine because `npos + 1 == 0` on many implementations, so the substring was "correctly" including the entire string. 

Thanks for pointing this out, I'll update this to be less implementation-dependent.

File: src/gennylib/src/quiesce.cpp
Comment (aggrand) [reply src/gennylib/src/quiesce.cpp]: Ended up making this a `warning` as per our offline discussion.

**Pull request merged**
